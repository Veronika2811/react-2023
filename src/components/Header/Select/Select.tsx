import { useRouter } from 'next/router';
import { ChangeEvent } from 'react';

import { DEFAULT_PAGE } from '@/utils/constants/constants';

import classes from './Select.module.css';

const Select = () => {
  const router = useRouter();

  const { name, perPage } = router.query;

  const onChangeItemsPerPage = (e: ChangeEvent<HTMLSelectElement>) => {
    const target = e.target.value;

    router.push({
      pathname: '/',
      query: {
        ...(name ? { name } : {}),
        page: DEFAULT_PAGE,
        perPage: target,
      },
    });
  };

  return (
    <select
      name="count-elements"
      defaultValue={perPage}
      onChange={onChangeItemsPerPage}
      className={classes.select}
      data-testid="select"
    >
      <option value="20" data-testid="option">
        20 cards
      </option>
      <option value="10" data-testid="option">
        10 cards
      </option>
    </select>
  );
};

export default Select;
