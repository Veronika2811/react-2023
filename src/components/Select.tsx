import { useContext } from 'react';
import { SetURLSearchParams } from 'react-router-dom';

import { CharactersContext } from '../context/context';

interface ISelectProps {
  setSearchParams: SetURLSearchParams;
}

const Select = ({ setSearchParams }: ISelectProps) => {
  const { setPerPage } = useContext(CharactersContext);

  return (
    <select
      className="select"
      name="count-elements"
      onChange={(e) => {
        setPerPage(+e.target.value);
        setSearchParams((searchParams) => {
          searchParams.set('page', '1');
          return searchParams;
        });
      }}
    >
      <option value="20">20 cards</option>
      <option value="10">10 cards</option>
    </select>
  );
};

export default Select;
