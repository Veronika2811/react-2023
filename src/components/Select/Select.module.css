.select {
  padding: 0.6em 1.2em;
  border: none;
  border-radius: 5px;
}