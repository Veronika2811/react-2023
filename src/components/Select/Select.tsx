import { ChangeEvent } from 'react';
// import { useSearchParams } from 'react-router-dom';

// import { useAppDispatch } from '@/store/hooks';
import { charactersChangePerPage } from '@/store/slice/charactersSlice';
import {
  // DEFAULT_PAGE,
  DEFAULT_VALUE_PER_PAGE,
  // PAGE_URL_PARAMETER_KEY,
} from '@/constants/constants';
import { useRouter } from 'next/router';

import classes from './Select.module.css';

const Select = () => {
  const router = useRouter()

  const { name, perPage, page } = router.query;

  // const [, setSearchParams] = useSearchParams();

  // const dispatch = useAppDispatch();

  const onChangeItemsPerPage = (e: ChangeEvent<HTMLSelectElement>) => {
    const target = e.target.value;

    router.push({
      pathname: '/',
      // query: { name: currentValue || '', page: page || 1 },
      query: { name: name || '', page: 1, perPage: +target || 20 },
      // query: currentValue? { name: currentValue} : {},
    })

    // dispatch(charactersChangePerPage(+target));

    // setSearchParams((searchParams) => {
    //   searchParams.set(PAGE_URL_PARAMETER_KEY, DEFAULT_PAGE);
    //   return searchParams;
    // });
  };

  return (
    <select
      name="count-elements"
      defaultValue={perPage}
      onChange={onChangeItemsPerPage}
      className={classes.select}
      data-testid="select"
    >
      <option value="20" data-testid="option">
        20 cards
      </option>
      <option value="10" data-testid="option">
        10 cards
      </option>
    </select>
  );
};

export default Select;
