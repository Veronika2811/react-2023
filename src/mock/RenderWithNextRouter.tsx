import React, { ReactNode } from 'react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import { createMockRouter } from './createMockRouter';

interface IRenderWithNextRouter {
  queryParams?: { [key: string]: string };
  children: ReactNode;
}

const RenderWithNextRouter = ({
  queryParams = {},
  children,
}: IRenderWithNextRouter) => {
  const { Provider } = RouterContext;

  console.log(children);

  return (
    <Provider value={createMockRouter({ query: queryParams })}>
      {children}
    </Provider>
  );
};

export default RenderWithNextRouter;
