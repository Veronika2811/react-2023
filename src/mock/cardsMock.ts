export const MortyMock = {
  created: '2017-11-04T18:50:21.651Z',
  episode: ['https://rickandmortyapi.com/api/episode/1'],
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  name: 'Morty Smith',
  origin: {
    name: 'unknown',
    url: '',
  },
  species: 'Human',
  status: 'Alive',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/2',
};

export const AlbertMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 11,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Albert Einstein',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
