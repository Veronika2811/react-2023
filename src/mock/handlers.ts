import { HttpResponse, http } from 'msw';

import { BASE_URL } from '@/utils/constants/constants';

export const handlers = [
  http.get(`${BASE_URL}`, () => {
    return HttpResponse.error();
  }),

  http.get(`${BASE_URL}/:id`, ({ params }) => {
    const { details } = params;
    return HttpResponse.json({
      created: '2017-11-04T18:50:21.651Z',
      episode: ['https://rickandmortyapi.com/api/episode/1'],
      gender: 'Male',
      id: details,
      image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
      location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
      },
      name: 'Morty Smith',
      origin: {
        name: 'unknown',
        url: '',
      },
      species: 'Human',
      status: 'Alive',
      type: '',
      url: 'https://rickandmortyapi.com/api/character/2',
    });
  }),
];