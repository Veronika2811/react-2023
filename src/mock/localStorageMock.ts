const mockLocalStorage = () => {
  const setItemMock = vi.fn();
  const getItemMock = vi.fn();

  beforeEach(() => {
    Storage.prototype.setItem = setItemMock;
    Storage.prototype.getItem = getItemMock;
  });

  afterEach(() => {
    setItemMock.mockRestore();
    getItemMock.mockRestore();
  });

  return { setItemMock, getItemMock };
};

export const { setItemMock, getItemMock } = mockLocalStorage();
