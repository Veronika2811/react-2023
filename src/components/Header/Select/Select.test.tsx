import { expect, it } from 'vitest';
import { screen, fireEvent, render } from '@testing-library/react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import Select from './Select';
import {
  ADDITIONAL_VALUE_PER_PAGE,
  DEFAULT_QUERY_PARAMS,
  DEFAULT_VALUE_PER_PAGE,
} from '@/utils/constants/constants';
import RenderWithNextRouter from '@/mock/RenderWithNextRouter';
import { createMockRouter } from '@/mock/createMockRouter';
import { AlbertMock } from '@/mock/cardsMock';

describe('Select component', () => {
  it('renders correctly Select component', () => {
    const container = render(
      <RenderWithNextRouter>
        <Select />
      </RenderWithNextRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should check the changes Select component', () => {
    render(
      <RenderWithNextRouter
        queryParams={{ perPage: DEFAULT_VALUE_PER_PAGE.toString() }}
      >
        <Select />
      </RenderWithNextRouter>
    );
    const options = screen.getAllByTestId('option') as HTMLOptionElement[];
    expect(options[0].selected).toBeTruthy();
    expect(options[1].selected).toBeFalsy();

    fireEvent.change(screen.getByTestId('select'), {
      target: { value: ADDITIONAL_VALUE_PER_PAGE },
    });

    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeTruthy();
  });

  it('should keep name from query params and go to 1 page when changing per page', () => {
    const router = createMockRouter({
      query: {
        name: AlbertMock.name,
        page: '5',
        perPage: DEFAULT_QUERY_PARAMS.perPage,
      },
    });

    render(
      <RouterContext.Provider value={router}>
        <Select />
      </RouterContext.Provider>
    );

    fireEvent.change(screen.getByTestId('select'), {
      target: { value: ADDITIONAL_VALUE_PER_PAGE },
    });

    expect(router.push).toBeCalledWith({
      pathname: '/',
      query: {
        page: '1',
        perPage: ADDITIONAL_VALUE_PER_PAGE,
        name: AlbertMock.name,
      },
    });
  });
});
