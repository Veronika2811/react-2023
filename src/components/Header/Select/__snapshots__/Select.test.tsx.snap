// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select component > renders correctly Select component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="_select_654fd1"
        data-testid="select"
        name="count-elements"
      >
        <option
          data-testid="option"
          value="20"
        >
          20 cards
        </option>
        <option
          data-testid="option"
          value="10"
        >
          10 cards
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="_select_654fd1"
      data-testid="select"
      name="count-elements"
    >
      <option
        data-testid="option"
        value="20"
      >
        20 cards
      </option>
      <option
        data-testid="option"
        value="10"
      >
        10 cards
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
